qiskit/circuit/quantumcircuitdata.py

```diff
@@ -14,9 +15,112 @@
+        mutations of the object do not invalidate the types, nor the restrictions placed on it by
+        its context.  Typically this will mean, for example, that :attr:`qubits` must be a sequence
+        of distinct items, with no duplicates.
+    """
+
+    __slots__ = ("operation", "qubits", "clbits", "_legacy_format_cache")
+
+    operation: Instruction
+    """The logical operation that this instruction represents an execution of."""
+    qubits: Tuple[Qubit, ...]
+    """A sequence of the qubits that the operation is applied to."""
+    clbits: Tuple[Clbit, ...]
+    """A sequence of the classical bits that this operation reads from or writes to."""
+
+    def __init__(
+        self,
+        operation: Instruction,
+        qubits: Iterable[Qubit] = (),
+        clbits: Iterable[Clbit] = (),
+    ):
+        self.operation = operation
+        self.qubits = tuple(qubits)
+        self.clbits = tuple(clbits)
+        self._legacy_format_cache = None
+
+    def copy(self) -> "CircuitInstruction":
+        """Return a shallow copy of the :class:`CircuitInstruction`."""
+        return self.__class__(
+            operation=self.operation,
+            qubits=self.qubits,
+            clbits=self.clbits,
+        )
+
+    def replace(
+        self,
+        operation: Optional[Instruction] = None,
+        qubits: Optional[Iterable[Qubit]] = None,
+        clbits: Optional[Iterable[Clbit]] = None,
+    ) -> "CircuitInstruction":
+        """Return a new :class:`CircuitInstruction` with the given fields replaced."""
+        return self.__class__(
+            operation=self.operation if operation is None else operation,
+            qubits=self.qubits if qubits is None else qubits,
+            clbits=self.clbits if clbits is None else clbits,
+        )
+
+    def __repr__(self):
+        return (
+            f"{type(self).__name__}("
+            f"operation={self.operation!r}"
+            f", qubits={self.qubits!r}"
+            f", clbits={self.clbits!r}"
+            ")"
+        )
+
+    def __eq__(self, other):
+        if isinstance(other, type(self)):
+            # Ordered from fastest comparisons to slowest.
+            return (
+                self.clbits == other.clbits
+                and self.qubits == other.qubits
+                and self.operation == other.operation
+            )
+        if isinstance(other, tuple):
+            return self._legacy_format == other
+        return NotImplemented
+
+    # Legacy tuple-like interface support.
+    #
+    # For a best attempt at API compatibility during the transition to using this new class, we need
+    # the interface to behave exactly like the old 3-tuple `(inst, qargs, cargs)` if it's treated
+    # like that via unpacking or similar.  That means that the `parameters` field is completely
+    # absent, and the qubits and clbits must be converted to lists.
+
+    @property
+    def _legacy_format(self):
+        if self._legacy_format_cache is None:
```

Comment:
We don't need to, because that would be an error in the old-style usage of the class - you can't assign to one component of a tuple (it's immutable).  I avoided doing it because it adds overhead in the new-style usage to support something that's an error in the old-style, and I wanted there to be zero overhead in new-style usage.